Mathlib/Tactic/TryThis.lean

```diff
@@ -32,21 +33,22 @@ def addHaveSuggestion (ref : Syntax) (t? : Option Expr) (e : Expr)
       `(tactic| let this := $estx)
   addSuggestion ref tac none origSpan?
 
-/-- Add a suggestion for `rw [h] at loc`. -/
-def addRewriteSuggestion (ref : Syntax) (e : Expr) (symm : Bool)
+open Lean.Parser.Tactic
+open Lean.Syntax
+
+/-- Add a suggestion for `rw [h₁, ← h₂] at loc`. -/
+def addRewriteSuggestion (ref : Syntax) (rules : List (Expr × Bool))
   (type? : Option Expr := none) (loc? : Option Expr := none)
   (origSpan? : Option Syntax := none) :
     TermElabM Unit := do
-  let estx ← delabToRefinableSyntax e
+  let rules_stx := TSepArray.ofElems <| ← rules.toArray.mapM fun ⟨e, symm⟩ => do
+    let t ← delabToRefinableSyntax e
+    if symm then `(rwRule| ← $t:term) else `(rwRule| $t:term)
   let tac ← do
-    let loc ← loc?.mapM fun loc => do `(Lean.Parser.Tactic.location| at $(← delab loc):term)
-    if symm then `(tactic| rw [← $estx] $(loc)?) else `(tactic| rw [$estx:term] $(loc)?)
+    let loc ← loc?.mapM fun loc => do `(location| at $(← delab loc):term)
+    `(tactic| rw [$rules_stx,*] $(loc)?)
 
-  let mut tacMsg :=
-    if let some loc := loc? then
-      if symm then m!"rw [← {e}] at {loc}" else m!"rw [{e}] at {loc}"
-    else
-      if symm then m!"rw [← {e}]" else m!"rw [{e}]"
+  let mut tacMsg := m!"{tac}"
```

Comment:
There seem to be some failing tests. I didn't spend any time debugging it, but there's a chance that all you need to do is use `addMessageContext` when you pass the massages into `addSuggestion`, since it seems like the problem is that the messages have forgotten the local context they were created in.

Reply:
Oh, nothing to do with the message: one of the `withMCtx` really needed to be a `setMCtx` (when we're in `!` mode).